File: src/com/colacelli/ircbot/plugins/joinpart/JoinPartPlugin.java
package com.colacelli.ircbot.plugins.joinpart;

import com.colacelli.ircbot.IRCBot;
import com.colacelli.ircbot.Plugin;
import com.colacelli.ircbot.listeners.OnChannelCommandListener;
import com.colacelli.ircbot.plugins.access.IRCBotAccess;
import com.colacelli.ircbot.plugins.help.PluginHelp;
import com.colacelli.ircbot.plugins.help.PluginHelper;
import com.colacelli.irclib.actors.Channel;
import com.colacelli.irclib.connection.Connection;
import com.colacelli.irclib.messages.ChannelMessage;

public class JoinPartPlugin implements Plugin {
    @Override
    public String getName() {
        return "JOIN_PART";
    }

    @Override
    public void onLoad(IRCBot bot) {
        IRCBotAccess.getInstance().addListener(bot, IRCBotAccess.OPERATOR_LEVEL, new OnChannelCommandListener() {
            @Override
            public String channelCommand() {
                return ".join";
            }

            @Override
            public void onChannelCommand(Connection connection, ChannelMessage message, String command, String... args) {
                if (args != null) {
                    connection.join(new Channel(args[0]));
                }
            }
        });
        PluginHelper.getInstance().addHelp(new PluginHelp(
                ".join",
                IRCBotAccess.OPERATOR_LEVEL,
                "Joins a channel",
                "#channel"));

        IRCBotAccess.getInstance().addListener(bot, IRCBotAccess.OPERATOR_LEVEL, new OnChannelCommandListener() {
            @Override
            public String channelCommand() {
                return ".part";
            }

            @Override
            public void onChannelCommand(Connection connection, ChannelMessage message, String command, String... args) {
                Channel channel = args == null ? message.getChannel() : new Channel(args[0]);
                connection.part(channel);
            }
        });
        PluginHelper.getInstance().addHelp(new PluginHelp(
                ".part",
                IRCBotAccess.OPERATOR_LEVEL,
                "Parts from a channel",
                "#channel"));
    }

    @Override
    public void onUnload(IRCBot bot) {
        String[] commands = {".join", ".part"};
        for (String command : commands) {
            PluginHelper.getInstance().removeHelp(command);
            IRCBotAccess.getInstance().removeListener(bot, command);
        }
    }
}

File: src/main/kotlin/com/colacelli/ircbot/plugins/joinpart/JoinPartPlugin.kt
package com.colacelli.ircbot.plugins.joinpart

import com.colacelli.ircbot.IRCBot
import com.colacelli.ircbot.base.Access
import com.colacelli.ircbot.base.Help
import com.colacelli.ircbot.base.Plugin
import com.colacelli.ircbot.base.listeners.OnChannelCommandListener
import com.colacelli.irclib.actors.Channel
import com.colacelli.irclib.actors.User
import com.colacelli.irclib.connection.Connection
import com.colacelli.irclib.connection.Server
import com.colacelli.irclib.connection.listeners.OnConnectListener
import com.colacelli.irclib.messages.ChannelMessage

class JoinPartPlugin : Plugin {
    val manager = ChannelsManager()

    override var name = "join_part"

    private var listener = object : OnConnectListener {
        override fun onConnect(connection: Connection, server: Server, user: User) {
            return manager.list().forEach {
                connection.join(Channel(it))
            }
        }
    }

    override fun onLoad(bot: IRCBot) {
        bot.addListener(listener)
        bot.addListener(object : OnChannelCommandListener {
            override val command = ".join"
            override val aliases: Nothing? = null
            override val level = Access.Level.OPERATOR
            override val help = Help(this, "Joins a channel", "#channel")

            override fun onChannelCommand(connection: Connection, message: ChannelMessage, command: String, args: Array<String>) {
                if (args.isNotEmpty()) {
                    manager.add(args[0])
                    connection.join(Channel(args[0]))
                }
            }

        })

        bot.addListener(object : OnChannelCommandListener {
            override val command = ".part"
            override val aliases: Nothing? = null
            override val level = Access.Level.OPERATOR
            override val help = Help(this, "Parts from a channel", "#channel")

            override fun onChannelCommand(connection: Connection, message: ChannelMessage, command: String, args: Array<String>) {
                if (args.isNotEmpty()) {
                    manager.del(args[0])
                    connection.part(Channel(args[0]))
                }
            }

        })
    }

    override fun onUnload(bot: IRCBot) {
        bot.addListener(listener)
        bot.removeListenersByCommands(arrayOf(".join", ".part"))
    }
}